src/web.rs

```diff
@@ -40,6 +40,26 @@ pub struct WebImpl {
     ctx: CanvasRenderingContext2d,
 }
 
+#[cfg(target_feature = "atomics")]
+#[wasm_bindgen]
+extern "C" {
+    #[wasm_bindgen(js_name = Uint8ClampedArray)]
+    type Uint8ClampedArray;
+
+    #[wasm_bindgen(constructor)]
+    fn new(array: &js_sys::Uint8Array) -> Uint8ClampedArray;
+}
+
+#[cfg(target_feature = "atomics")]
+#[wasm_bindgen]
+extern "C" {
+    #[wasm_bindgen(js_name = ImageData)]
+    type ImageDataExt;
+
+    #[wasm_bindgen(catch, constructor, js_class = "ImageData")]
+    fn new(array: Uint8ClampedArray, sw: u32) -> Result<ImageDataExt, wasm_bindgen::JsValue>;
+}
+
```

Comment:
Is this not available in `wasm-bindgen` yet?

Reply:
Looking at my own comment (https://github.com/rust-windowing/softbuffer/pull/77#issuecomment-1450402871) again, it actually didn't explain why this was necessary and I barely remembered anything myself. Apologies.

There is no `Uint8ClampedArray` in `web-sys`, there is [`Clamped`](https://docs.rs/wasm-bindgen/0.2.84/wasm_bindgen/struct.Clamped.html), which is supposed to fulfill the role of `Uint8ClampedArray`. I am assuming this was done for good reason: to implement the special behavior that is necessary to prevent copies when using it with slices.

The issue is that `Clamped` takes a generic, which can be a `Vec<u8>`, which presumably does what we want here, a copy, but `wasm-bindgen` doesn't support generics in their exports or imports. So, I presume, the decision was made to ignore the generic and only take `Clamped<&[u8]>`, which should be what you always want anyway.

Unfortunately, `ImageData` doesn't take a `Uint8ClampedArray` based on a `SharedArray`, which is what the underlying memory of `wasm-bindgen` is when you use the `atomics` target feature, and is what `Clamped` uses under the hood. This is why it will simply throw when using the `atomics` target feature. This is why a copy is necessary.

I checked again if this can somehow be upstreamed. The obvious solution is to implement generics, which is a huge project and probably doesn't have the reviewer capacity for it anyway.
The other option would be to adjust the WebIDL generator to include multiple methods to cover all these cases, but that would be very ugly and adds weird names, it's weird enough already, I will ask though.

I want to stress, though, that there is nothing wrong with adding custom `wasm-bindgen` definitions, it's the same mechanic `web-sys` uses too. It is frequently done by many projects to circumvent weird problems like this.

Reply:
Ha, this is funny, I just started researching how complicated a fix would really be and found that `wasm-bindgen` itself is using the same workaround in their own example.

But (!), there's is simpler, because somehow I overlooked that `Uint8ClampedArray` is actually implemented, but in `js-sys`. Don't know how I overlooked this.

See https://github.com/rustwasm/wasm-bindgen/blob/e7c3f52e15f73702d8c94959574bcee1a0d86227/examples/raytrace-parallel/src/lib.rs#L119-L125.

I just pushed a fix for that.